### ISSUE TITLE
CA1816 shouldn't warn for IAsyncDisposable.DisposeAsync

### ISSUE THREAD

**Issue by kevinoid** [dotnet/roslyn-analyzers#3675]
#### Analyzer package

[Microsoft.NetCore.Analyzers](https://www.nuget.org/packages/Microsoft.NetCore.Analyzers/)

#### Package Version

v3.0.0 (Latest)

#### Diagnostic ID

[CA1816](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1816)

#### Repro steps

Create a class which implements [`IAsyncDisposable`](https://docs.microsoft.com/en-us/dotnet/api/system.iasyncdisposable) and calls [`GC.SuppressFinalize`](https://docs.microsoft.com/en-us/dotnet/api/system.gc.suppressfinalize) in [`DisposeAsync`](https://docs.microsoft.com/en-us/dotnet/api/system.iasyncdisposable.disposeasync).  For example:

```csharp
namespace CA1816Example
{
    using System;
    using System.Runtime.InteropServices;
    using System.Threading.Tasks;

    class MyAsyncDisposable : IAsyncDisposable
    {
        [DllImport("example.dll")]
        private static extern int GetHandle();

        [DllImport("example.dll")]
        private static extern void FreeHandle(int handle);

        private readonly int handle;

        public MyAsyncDisposable()
        {
            this.handle = GetHandle();
        }

        ~MyAsyncDisposable()
        {
            FreeHandle(this.handle);
        }

        public async ValueTask DisposeAsync()
        {
            await Task.Run(() => FreeHandle(this.handle)).ConfigureAwait(false);
            GC.SuppressFinalize(this);
        }
    }
}
```

#### Expected behavior

CA1816 would not be produced.

#### Actual behavior

    MyAsyncDisposable.cs(30,13): warning CA1816: MyAsyncDisposable.DisposeAsync() calls GC.SuppressFinalize(object), a method that is typically only called within an implementation of 'IDisposable.Dispose'. Refer to the IDisposable pattern for more information.

#### Discussion

Although I haven't seen official guidance for [dispose pattern](https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose#implement-the-dispose-pattern) with `IAsyncDisposable`, it seems likely that classes which define a finalizer (e.g. to release native resources) would want to suppress finalization if the work has been completed by `DisposeAsync`, as in the above example, and therefore CA1816 shouldn't apply to use in this method.  Thoughts?

**Comment by kevinoid** [comment]
For an example in the framework, see [`XmlWriterAsync`](https://github.com/dotnet/runtime/blob/77fab0b/src/libraries/System.Private.Xml/src/System/Xml/Core/XmlWriterAsync.cs#L607).

**Comment by mavasani** [comment]
@Evangelink similar to the issue you recently handled for CA2213...

**Comment by skrysmanski** [comment]
Problem still exists in 3.3.

Also, in the meantime there's now an official guidance on how to implement `IAsyncDisposable` (which contains a call to `GC.SuppressFinalize()`): <https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-disposeasync>

**Comment by Evangelink** [comment]
@skrysmanski Yes you are right but there is an open PR fixing the issue, it is just waiting for approval and merge :)
